### ISSUE TITLE
NameError: uninitialized constant RSpec::Core::Version

### ISSUE THREAD

**Issue by sj26** [KnapsackPro/knapsack_pro-ruby#150]
I found an error while setting up a reproduction environment. I did a fresh setup of knapsack pro in an rspec environment. It was running okay:
https://buildkite.com/sj26/knapsack-test/builds/6

Then I added `KNAPSACK_PRO_RSPEC_SPLIT_BY_TEST_EXAMPLES=true` and I got an error:
https://buildkite.com/sj26/knapsack-test/builds/8#15887146-f8e4-458a-9255-b6c999cb49d6/79-83

```
** Invoke knapsack_pro:rspec (first_time)
** Execute knapsack_pro:rspec
rake aborted!
NameError: uninitialized constant RSpec::Core::Version
/opt/rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/knapsack_pro-2.18.0/lib/knapsack_pro/base_allocator_builder.rb:37:in `fast_and_slow_test_files_to_run'
/opt/rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/knapsack_pro-2.18.0/lib/knapsack_pro/allocator_builder.rb:5:in `allocator'
/opt/rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/knapsack_pro-2.18.0/lib/knapsack_pro/runners/base_runner.rb:10:in `initialize'
/opt/rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/knapsack_pro-2.18.0/lib/knapsack_pro/runners/rspec_runner.rb:9:in `new'
/opt/rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/knapsack_pro-2.18.0/lib/knapsack_pro/runners/rspec_runner.rb:9:in `run'
/opt/rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/knapsack_pro-2.18.0/lib/tasks/rspec.rake:5:in `block (2 levels) in <main>'
/opt/rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/rake-13.0.6/lib/rake/task.rb:279:in `block in execute'
/opt/rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/rake-13.0.6/lib/rake/task.rb:279:in `each'
/opt/rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/rake-13.0.6/lib/rake/task.rb:279:in `execute'
/opt/rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/rake-13.0.6/lib/rake/task.rb:219:in `block in invoke_with_call_chain'
/opt/rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/rake-13.0.6/lib/rake/task.rb:199:in `synchronize'
/opt/rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/rake-13.0.6/lib/rake/task.rb:199:in `invoke_with_call_chain'
/opt/rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/rake-13.0.6/lib/rake/task.rb:188:in `invoke'
/opt/rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/rake-13.0.6/lib/rake/application.rb:160:in `invoke_task'
/opt/rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/rake-13.0.6/lib/rake/application.rb:116:in `block (2 levels) in top_level'
/opt/rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/rake-13.0.6/lib/rake/application.rb:116:in `each'
/opt/rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/rake-13.0.6/lib/rake/application.rb:116:in `block in top_level'
/opt/rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/rake-13.0.6/lib/rake/application.rb:125:in `run_with_threads'
/opt/rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/rake-13.0.6/lib/rake/application.rb:110:in `top_level'
/opt/rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/rake-13.0.6/lib/rake/application.rb:83:in `block in run'
/opt/rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/rake-13.0.6/lib/rake/application.rb:186:in `standard_exception_handling'
/opt/rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/rake-13.0.6/lib/rake/application.rb:80:in `run'
bin/rake:5:in `<main>'
Tasks: TOP => knapsack_pro:rspec
```

Here's the full source code at that point:
https://github.com/sj26/knapsack-test/tree/rspec-core-version-error

The command being run is in the pipeline file:
https://github.com/sj26/knapsack-test/blob/rspec-core-version-error/.buildkite/pipeline.yml#L23

There might be a `require` missing somewhere?

**Comment by sj26** [comment]
A workaround is to insert a rake task dependency which requires the file:
https://github.com/sj26/knapsack-test/compare/rspec-core-version-error...rspec-core-version-workaround

**Comment by ArturT** [comment]
Most likely you have not set RAILS_ENV=test and rspec is not loaded.

**Comment by ArturT** [comment]
You can also ensure RSpec in `test` and `development` group in your Gemfile.

**Comment by sj26** [comment]
RAILS_ENV is set to test:
https://github.com/sj26/knapsack-test/blob/rspec-core-version-error/.buildkite/pipeline.yml#L16

rspec is in the test and development groups:
https://github.com/sj26/knapsack-test/blob/rspec-core-version-error/Gemfile#L8-L14

Are you expecting rails rake to have called `Bundler.require`? It doesn't, on purpose, so that invoking rake is quick.

**Comment by ArturT** [comment]
That's interesting. You basically use knapsack_pro in a way that's not the default approach documented in README. Most people just use bundler:

```
bundle exec rake knapsack_pro:rspec
```

You use `bin/rake` to load stuff faster and load Rails environment only if RSpec config (like spec_helper.rb or rails_helper.rb) decides to do it. I could imagine someone could have tests that don't depend on Rails environment and then the user doesn't have to load Rails environment so using `bin/rake` makes more sense than `bundle exec` because `bin/rake` is faster.

Anyway. I didn't think about it before so I never covered the `bin/rake` approach in the documentation that's why you encounter the issue.

**Comment by ArturT** [comment]
Maybe adding `require 'rspec/core/version'` inside of the method `fast_and_slow_test_files_to_run` right before `RSpec::Core::Version::STRING` check would be enough to fix the issue. I need to ensure we load rspec only when someone is running RSpec test runner.

https://github.com/KnapsackPro/knapsack_pro-ruby/blob/4e0c5c18e6534e3ad1bafe1b9f49c9fb538614d7/lib/knapsack_pro/base_allocator_builder.rb#L37

**Comment by sj26** [comment]
`bundle exec rake` is roughly the same as `bin/rake`, and has the same error:

```
RAILS_ENV=test KNAPSACK_PRO_RSPEC_SPLIT_BY_TEST_EXAMPLES=true bundle exec rake knapsack_pro:rspec                       
rake aborted!
NameError: uninitialized constant RSpec::Core::Version
/opt/rbenv/versions/2.7.2/bin/bundle:23:in `load'
/opt/rbenv/versions/2.7.2/bin/bundle:23:in `<main>'
Tasks: TOP => knapsack_pro:rspec
(See full trace by running task with --trace)
```

> Maybe adding require 'rspec/core/version' inside of the method fast_and_slow_test_files_to_run right before RSpec::Core::Version::STRING check would be enough to fix the issue.

Yeah, that's what I was thinking! :+1:

**Comment by sj26** [comment]
Another alternative which avoids requiring the file would be to look at the loaded spec version, which `bundle exec ...` sets up:

```ruby
      if adapter_class == KnapsackPro::Adapters::RSpecAdapter && KnapsackPro::Config::Env.rspec_split_by_test_examples?
        unless Gem.loaded_specs["rspec-core"].version >= Gem::Version.new('3.3.0')
```

**Comment by ArturT** [comment]
I've released knapsack_pro 2.18.1 with a fix. 
PR: https://github.com/KnapsackPro/knapsack_pro-ruby/pull/151

**Comment by sj26** [comment]
Thanks @ArturT!
